telemetry/telemetry.go

```diff
@@ -0,0 +1,111 @@
+// As part of the White Glove White Glove project,
+// this is a telemetry package to support posting summary and cancellation events
+// to a given webhook endpoint with a reference project id in the shape of an event grid
+// custom event.
+
+package telemetry
+
+import (
+	"bytes"
+	"encoding/json"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/Azure/azure-storage-azcopy/common"
+)
+
+var (
+	// ReportingAPIURI the Webhook URI
+	ReportingAPIURI string
+	// ProjectReferenceID the project reference ID
+	ProjectReferenceID string
+)
+
+type telemetry struct {
+	JobID     string      `json:"jobID"`
+	ProjectID string      `json:"projectId"`
+	Telemetry interface{} `json:"telemetry"`
+}
+
+type reportEvent struct {
+	EventType  string      `json:"eventType"`
+	EventTopic string      `json:"eventTopic"`
+	Data       interface{} `json:"data"`
+	EventTime  time.Time   `json:"eventTime"`
+}
+
+func getJSON(t interface{}) string {
+	json, err := json.Marshal(t)
+	common.PanicIfErr(err)
+	return string(json)
+}
+
+func newReportEvent(eventType string, data interface{}) *reportEvent {
+	return &reportEvent{
+		EventType:  eventType,
+		EventTopic: "",
+		Data:       data,
+		EventTime:  time.Now(),
+	}
+}
+
+func postEvent(json string) {
+	var jsonStr = []byte(json)
+	_, err := http.Post(ReportingAPIURI, "application/json", bytes.NewBuffer(jsonStr))
```

Comment:
It's important to re-use the same HTTP client object for all calls here, since it is the httpClient class that does connection pooling in Go.  See one of the comments on this commit, for details: https://github.com/Azure/azure-storage-azcopy/commit/fad439f75b7860797f4d482b3e378995007f7b4f